src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1500,14 +1530,225 @@ public Durabilty getDurabilty() {
+        // This absolutely must come first, or else reading the checksum becomes part of the checksum
+        long expectedChecksum = in.getChecksum();
+        long readChecksum = in.readInt() & 0xFFFF_FFFFL;
+        if (readChecksum != expectedChecksum) {
+            throw new TranslogCorruptedException("translog stream is corrupted, expected: 0x" +
+                    Long.toHexString(expectedChecksum) + ", got: 0x" + Long.toHexString(readChecksum));
+        }
+    }
+
+    public static Snapshot snapshotFromStream(StreamInput input, final int numOps) {
+        final BufferedChecksumStreamInput checksumStreamInput = new BufferedChecksumStreamInput(input);
+        return new Snapshot() {
+            int read = 0;
+            @Override
+            public int estimatedTotalOperations() {
+                return numOps;
+            }
+
+            @Override
+            public Operation next() throws IOException {
+                if (read < numOps) {
+                    read++;
+                    return readOperation(checksumStreamInput);
+                }
+                return null;
+            }
+
+            @Override
+            public void close() {
+                // doNothing
+            }
+        };
+    }
+
+    static Translog.Operation readOperation(BufferedChecksumStreamInput in) throws IOException {
+        int opSize = in.readInt();
+        Translog.Operation operation;
+        try {
+            in.resetDigest(); // size is not part of the checksum?
+            if (in.markSupported()) { // if we can we validate the checksum first
+                in.mark(opSize);
+                if (opSize < 4) { // 4byte for the checksum
+                    throw new TranslogCorruptedException("operation size must be at least 4 but was: " + opSize);
+                }
+                in.skip(opSize-4);
+                verifyChecksum(in);
+                in.reset();
+            }
+            Translog.Operation.Type type = Translog.Operation.Type.fromId(in.readByte());
+            operation = newOperationFromType(type);
+            operation.readFrom(in);
+            verifyChecksum(in);
+        } catch (EOFException e) {
+            throw new TruncatedTranslogException("reached premature end of file, translog is truncated", e);
+        } catch (AssertionError|Exception e) {
+            throw new TranslogCorruptedException("translog corruption while reading from stream", e);
+        }
+        return operation;
+    }
+
+    public static void writeOperation(StreamOutput outStream, Translog.Operation op) throws IOException {
+        // We first write to a NoopStreamOutput to get the size of the
+        // operation. We could write to a byte array and then send that as an
+        // alternative, but here we choose to use CPU over allocating new
+        // byte arrays.
+        NoopStreamOutput noopOut = new NoopStreamOutput();
```

Comment:
Hmm this double-write seems crazy ... the comment is wrong I think: we are already recycling byte[] (using BigArrays under this StreamOutput)?  Seems like TranslogWriter.add could handle writeInt(size), so this code just needs to write the operation & its checksum?

Anyway, we should fix this separately/later ... it's pre-existing, and it's just an optimization.